src/cast_core/include/cast_core/actors/OptionsConversion.hpp

```diff
@@ -118,37 +125,277 @@ struct NodeConvert<mongocxx::options::count> {
+        case YAML::NodeType::Undefined:
+        case YAML::NodeType::Scalar:
+        case YAML::NodeType::Null: {
+            Value val = Value::parseScalar(node);
+            val.appendToBuilder(doc, key);
+            return;
+        }
+    }
+    std::abort();
+}
+
+void appendToBuilder(const YAML::Node& node, bsoncxx::builder::basic::array& arr) {
+    switch (node.Type()) {
+        case YAML::NodeType::Map:
+            arr.append(toDocumentBson(node));
+            return;
+        case YAML::NodeType::Sequence:
+            arr.append(toArrayBson(node));
+            return;
+        case YAML::NodeType::Undefined:
+        case YAML::NodeType::Scalar:
+        case YAML::NodeType::Null: {
+            Value val = Value::parseScalar(node);
+            val.appendToBuilder(arr);
+            return;
+        }
+    }
+
+    std::abort();
+}
+
+bsoncxx::document::value toDocumentBson(const YAML::Node& node) {
+    if (!node.IsMap()) {
+        std::stringstream msg;
+        msg << "Wanted map got " << node.Type() << ": " << to_string(node);
+        BOOST_THROW_EXCEPTION(InvalidYAMLToBsonException(msg.str()));
+    }
+
+    bsoncxx::builder::basic::document doc{};
+    for (auto&& kvp : node) {
+        appendToBuilder(kvp.second, kvp.first.as<std::string>(), doc);
+    }
+    return doc.extract();
+}
+
+bsoncxx::array::value toArrayBson(const YAML::Node& node) {
+    if (!node.IsSequence()) {
+        std::stringstream msg;
+        msg << "Wanted sequence got " << node.Type() << ": " << to_string(node);
+        BOOST_THROW_EXCEPTION(InvalidYAMLToBsonException(msg.str()));
+    }
+
+    bsoncxx::builder::basic::array arr{};
+    for (auto&& elt : node) {
+        appendToBuilder(elt, arr);
+    }
+    return arr.extract();
+}
+
+bsoncxx::document::value toDocumentBson(const genny::Node& node) {
+    std::stringstream str;
+    str << node;
+    const YAML::Node asYaml = YAML::Load(str.str());
+    return genny::toDocumentBson(asYaml);
+}
+
+bsoncxx::document::value toDocumentBson(const std::string& yaml) {
+    const YAML::Node asYaml = YAML::Load(yaml);
+    return genny::toDocumentBson(asYaml);
+}
+
+bsoncxx::array::value toArrayBson(const std::string& yaml) {
+    const YAML::Node asYaml = YAML::Load(yaml);
+    return genny::toArrayBson(asYaml);
+}
+
+}  // namespace
+
+template <>
+struct NodeConvert<bsoncxx::document::value> {
+    using type = bsoncxx::document::value;
+
+    static type convert(const Node& node) {
+        return toDocumentBson(node);
+    }
+};
+
 template <>
 struct NodeConvert<mongocxx::options::find> {
     using type = mongocxx::options::find;
 
     static type convert(const Node& node) {
         type rhs{};
 
-        if (node["Hint"]) {
-            auto h = node["Hint"].to<std::string>();
-            auto hint = mongocxx::hint(std::move(h));
-            rhs.hint(mongocxx::hint(hint));
+        if (const auto& sort = node["Sort"]) {
+            rhs.sort(sort.to<bsoncxx::document::value>());
```

Comment:
The conversions from yaml to C++ driver structs for options and the like has always bothered me. Ultimately it all just gets sent as bson; the type-system just gets in the way in this kind of library/framework code.

(Aside: One advantage of hand-crafted conversions is we can have intuitive error-messages and guard against mis-configurations (e.g. projections on empty finds).)

I'd _love_ to just use extended json. Perhaps there's 3p code that we could crib. But ultimately don't we still have to map to C++ driver structs since that's the driver's API?

I don't have a great suggestion for how to handle this. We haven't prioritized any foundational changes like this in Genny for a while, so I'm hesitant to suggest we ticket that work out. Perhaps add it to the [genny ideas doc](https://docs.google.com/document/d/1K0SlHoKkIACIXhkNP_KPHvb3IV6r53ReqsHqEYTnXDA/edit#heading=h.9sx1dtlnczfk) which we intend to 💅 once we ramp up on more genny investment in the (hopefully near) future.

Reply:
On this note: now that the genny test code isn't being shared anymore, I'm going to close  https://jira.mongodb.org/browse/EVG-18364 as gone away unless there's other refactoring that can be done

Reply:
Woo! I'm happy that this was easy to do. I'll leave the thread open in case @rtimmons has anything to add, but I'm happy with its resolution.